### PR TITLE
feat: Add support for `TorchTensorRTModule` in Dynamo [1 / x]

### LINKED ISSUE #2019: ✨[Feature] Toggle usage of `TRTModuleNext` based on C++ availability
# Description
Based on [this comment](https://github.com/pytorch/TensorRT/pull/2003#discussion_r1228697019) by @narendasan, the usage of the `TRTModuleNext` functionality, now named `TorchTensorRTModule`, could be automatically toggled based on the availability of the C++ functionality, like so:
```python
try: 
    import TorchTensorRTModule
    USE_PYTHON_RT = False
except:
    USE_PYTHON_RT = True
```

# Related Issues
#1993, #1943

### PR DESCRIPTION
**Pull request by gs-olive**
# Description

- Rename `TRTModuleNext` to `TorchTensorRTModule` across the repository, and move the source directory to `dynamo`
- Update imports across the repository
- Refactor `convert_module` code to support conversion to a `TorchTensorRTModule`
- Add tests for `TorchTensorRTModule` functionality in Dynamo

Fixes #2019 

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ x ] I have added tests to verify my fix or my feature
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/__init__.py [hunk 1]
```diff
@@ -91,7 +91,7 @@ def _find_lib(name, paths):
 from torch_tensorrt import logging
 from torch_tensorrt._Input import Input
 from torch_tensorrt._Device import Device
-from torch_tensorrt._TRTModuleNext import TRTModuleNext
+from torch_tensorrt.dynamo import TorchTensorRTModule
```
Comment (gs-olive) [inline py/torch_tensorrt/__init__.py:94]: This gives users the option to _either_ do `from torch_tensorrt.dynamo import TorchTensorRTModule` or `from torch_tensorrt import TorchTensorRTModule`. We can remove the latter by removing this line.

Comment (gs-olive) [reply py/torch_tensorrt/__init__.py:94]: Causes a circular import issue - removing this line in favor of `from torch_tensorrt.dynamo import TorchTensorRTModule`

File: py/torch_tensorrt/dynamo/backend/__init__.py [hunk 2]
```diff
@@ -45,6 +46,7 @@ def compile(
     min_block_size=MIN_BLOCK_SIZE,
     torch_executed_ops=[],
     torch_executed_modules=[],
+    use_experimental_rt=USE_EXPERIMENTAL_RT,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/backend/__init__.py:49]: Can you create an issue to flip this to use_python_rt?
The behavior can be like in defaults:
```py
try: 
    import TorchTensorRTModule
    USE_PYTHON_RT = False
except:
    USE_PYTHON_RT = True
```

File: py/torch_tensorrt/dynamo/backend/__init__.py [hunk 3]
```diff
@@ -57,6 +59,13 @@ def compile(
         + "torch_executed_ops, pass_through_build_failures}"
     )
 
+    if "use_experimental_fx_rt" in kwargs:
+        logger.info(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/backend/__init__.py:63]: Dont think you need to log this. If anything just print out what runtime is being used

File: py/torch_tensorrt/dynamo/backend/__init__.py [hunk 4]
```diff
@@ -57,6 +65,11 @@ def compile(
         + "torch_executed_ops, pass_through_build_failures}"
     )
 
+    if "use_experimental_fx_rt" in kwargs:
+        use_experimental_rt = kwargs["use_experimental_fx_rt"]
+
+    logger.info(f"Using {'C++' if use_experimental_rt else 'Python'} TRT Runtime")
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/backend/__init__.py:71]: Reporting which runtime is being used here, to the user

File: py/torch_tensorrt/dynamo/__init__.py [hunk 5]
```diff
@@ -1,2 +1,3 @@
+from ._TorchTensorRTModule import TorchTensorRTModule
```
[the remark this answers is not in the document]
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/__init__.py]: Resolved by moving the import of `TorchTensorRTModule` directly where the module is invoked.

File: py/torch_tensorrt/dynamo/fx_ts_compat/lower.py [hunk 6]
```diff
@@ -70,7 +71,9 @@ def compile(
         timing_cache_prefix: Timing cache file name for timing cache used by fx2trt.
         save_timing_cache: Update timing cache with current timing cache data if set to True.
         cuda_graph_batch_size: Cuda graph batch size, default to be -1.
-        use_experimental_fx_rt: Uses the next generation TRTModule which supports both Python and TorchScript based execution (including in C++).
+        use_python_runtime: Whether to strictly use Python runtime or C++ runtime. To auto-select a runtime
+            based on C++ dependency presence (preferentially choosing C++ runtime if available), leave the
+            argument as None
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/fx_ts_compat/lower.py:76]: Added a new compilation argument called `use_python_runtime` to replace `use_experimental_rt` and `use_experimental_fx_rt`. This new compilation argument has three potential settings/implications:

- `True`: Strictly use Python runtime
- `False`: Strictly use C++ runtime
- `None` [**DEFAULT**]: Automatically detect presence of C++ dependency and select C++ runtime if available, else default to Python runtime

File: py/torch_tensorrt/dynamo/test/test_dynamo_backend.py [hunk 7]
```diff
@@ -31,6 +31,8 @@ def test_resnet18(ir):
         "enabled_precisions": {torch.float},
         "ir": ir,
         "pass_through_build_failures": True,
+        "optimization_level": 1,
+        "min_block_size": 8,
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/test/test_dynamo_backend.py:35]: Decreased optimization level (from 3 to 1) and increased block size (from 5 to 8) to improve compilation times and reduce segmentation in models, pending upcoming converter implementations.

**Pull request merged**